### ISSUE TITLE
RDP session closes when the session lock screen appears. Seems related to the recent clipboard updates.

### ISSUE THREAD

**Issue by Tiertier** [neutrinolabs/xrdp#1955]
Hello XRDP development team,

I currently use XRDP to connect to existing VNC sessions. So I encountered the copy/paste problem for messages that close RDP session.
While browsing the XRDP GitHub, I saw that this issue has been fixed and reviewed recently but there is no official release yet.
So, I downloaded the available source code and I made the installation from the source following the instructions described.

At first sight, the copy and paste function seems to work correctly from the Windows built-in RDP client. 

However, this update seems to release a new problem.
Every time my session locks after a few minutes of inactivity or I lock my session manually. As soon as I move the mouse back into the window to access it again, it closes instantly.

In response to this event, I found this in the logfile:

[*DATE*-*TIME*] [DEBUG] VNC got clip data
[*DATE*-*TIME*] [ERROR] Memory exhausted allocating 0 bytes for RFB clip data
[*DATE*-*TIME*] [ERROR] lib_data_in: lib_mod_process_message failed
[*DATE*-*TIME*] [DEBUG] Closed socket 12 (AF_INET **RDP_SERVER_IP**:**PORT**)
[*DATE*-*TIME*] [DEBUG] xrdp_mm_module_cleanup
[*DATE*-*TIME*] [DEBUG] Closed socket 18 (AF_INET **VNC_SERVER_IP**:**PORT**)

Would you have a solution to solve this problem?

I can also share any additional information needed to debug this issue.

Thanks !

**Comment by matt335672** [comment]
Hi @Tiertier 

Thanks very much for reporting this so clearly. It looks like a programming error on my part (and a fairly basic one at that, somewhat embarrassingly). On the plus side we should be able to fix it pretty quickly.

Are you in a position to try this patch on the source code you downloaded?

[clip.patch](https://github.com/neutrinolabs/xrdp/files/6908076/clip.patch.txt)

Download it to your build machine (say in `/tmp/clip.patch`), change to your build directory, and apply with `patch -p1 </tmp/clip.patch`.

I haven't tested this myself yet, as I'm about to finish for the weekend.

Thanks.

**Comment by Tiertier** [comment]
Hi @matt335672

Thank you very much for this reactivity !

I just applied the patch provided above. The problem seems to be well solved according to my first tests.

The clipboard also seems to work perfectly and in both directions (with version 0.9.6, it only worked from the remote session to the client). Also when I was copying long messages, my session was instantly closed by the server. This doesn't seem to happen anymore either. 

During the first tests, I noticed that when the message is too long. If I copy from the client to the remote session, the message is copied but shortened to what I assume is the copy buffer size limit. On the other hand, in the remote session to client direction, the data is copied within the remote session but is not transmitted to the client. I don't know if this is the normal behavior it should adopt?

I will continue to use this more in depth over the weekend and next week to do my stuff. This will allow me to give you my user feedback.

Anyway, thanks for these improvements, it will already greatly improve my user experience.

**Comment by matt335672** [comment]
Thanks for the quick feedback.

I didn't think we had any copy buffer limits, so if you can give me an example I can look at, that would help. I'll happily put some more patches together for you. We've got a release at the end of August, so it's a great time to get user-reported fixes in.

**Comment by Tiertier** [comment]
Hello,

I have just done some repetitive and more precise tests to explain it to you and it seems that I was too fast in writing my previous message. 

The behavior looks the same in both directions when copying a too long message. Here is what I did as a test:

I simply opened a text file with gedit in which I typed several times a line containing 100 characters each as below: 
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa

If the file contains 2000 lines (100000 characters), the copy is correct in both directions.
If the file contains 3000 lines (300000 characters), the copy is not transmitted.

This behavior is the same whether you use xfreerdp or the built-in Microsoft RDP client.

 I assume that the clipboard has a maximum size which should be something like 256Kb. This doesn't sound like a problem in the end, a limit between 200.000 and 300.000 characters seems to me very correct !

**Comment by matt335672** [comment]
Hi @Tiertier 

Thanks again for the clear and lucid report.

May I ask if you are using tigervnc? If so, read the description of `-MaxCutText` [here](https://tigervnc.org/doc/Xvnc.html)

If you want to play with this setting, you can pass it in to Xvnc via sesman.ini. Let me know how it goes.

**Comment by matt335672** [comment]
Hi @Tiertier 

Have you got an update on this? Sorry to hassle you, but I'd like to get the patch above into devel ASAP if it's actually complete.

Thanks.

**Comment by Tiertier** [comment]
Hi @matt335672,

Thank you for your quick answers and help. Sorry for the late answer ...

I am indeed using TigerVNC, I was able to identify that the clipboard limit was coming from the default configuration of TigerVNC and not from XRDP.

To be honest, I still have some clipboard problems with the xfreeRDP client (version 2.2.0). The clipboard doesn't work anymore at some time in the client to remote session direction. But I haven't had time to investigate to identify the recurrence pattern yet. Also, I saw that xfreeRDP recently released a new version with clipboard fixes as well, I will install and test it.
But these problems don't appear when I use the Windows built-in client, so the problem should not come from the XRDP server.

I can keep you posted on this, but from my perspective, the patch is fully functional.

Thank you again.
